File: nodejs/packages/periph/src/connection/neo6_mock.js
'use strict';

/**
 * In-memory fake connection for NEO-6 unit tests — no hardware, no bus.
 *
 * NEO-6's driver treats UART, I2C (DDC), and SPI as the same underlying
 * NMEA/UBX byte stream (see specs/gnss/neo-6.md's "Connection abstraction"
 * note): `read(1)` for UART, `writeRead(Buffer.from([0xFF]), 1)` for I2C,
 * `writeRead(Buffer.alloc(0), 1)` for SPI. This mock is transport-shape
 * agnostic: preload the stream with `queueBytes(data)`; `read()` and
 * `writeRead()` both just pop the next byte(s) off the front of one shared
 * FIFO, regardless of the prefix passed to `writeRead` — matching the real
 * module, where all three transports deliver the same bytes and only the
 * framing differs. `write()` calls (e.g. sendUbx) are logged to `writes`
 * for assertions.
 */
class Neo6ConnectionMock {
    constructor() {
        this._stream = Buffer.alloc(0);
        this.writes = [];
    }

    /** Append bytes to the end of the shared read stream. */
    queueBytes(data) {
        this._stream = Buffer.concat([this._stream, Buffer.from(data)]);
    }

    async read(n) {
        const out = this._stream.subarray(0, n);
        this._stream = this._stream.subarray(out.length);
        return Buffer.from(out);
    }

    async writeRead(data, n) {
        return this.read(n);
    }

    async write(data) {
        this.writes.push(Buffer.from(data));
    }

    async close() {}
}

module.exports = { Neo6ConnectionMock };

File: python/periph/connection/neo6_mock.py
class NEO6ConnectionMock:
    """In-memory fake connection for NEO-6 unit tests — no hardware, no bus.

    NEO-6's driver treats UART, I2C (DDC), and SPI as the same underlying
    NMEA/UBX byte stream (see specs/gnss/neo-6.md's "Connection
    abstraction" note): ``read(1)`` for UART, ``write_read(b'\\xff', 1)``
    for I2C, ``write_read(b'', 1)`` for SPI. This mock is transport-shape
    agnostic: preload the stream with ``queue_bytes(data)``; ``read()`` and
    ``write_read()`` both just pop the next byte(s) off the front of one
    shared FIFO, regardless of the prefix passed to ``write_read`` —
    matching the real module, where all three transports deliver the same
    bytes and only the framing differs. ``write()`` calls (e.g. send_ubx)
    are logged to ``writes`` for assertions.
    """

    def __init__(self):
        self._stream = bytearray()
        self.writes = []

    def queue_bytes(self, data):
        """Append bytes to the end of the shared read stream."""
        self._stream.extend(data)

    def read(self, n):
        out = bytes(self._stream[:n])
        del self._stream[:len(out)]
        return out

    def write_read(self, data, n):
        return self.read(n)

    def write(self, data):
        self.writes.append(bytes(data))

    def close(self):
        pass
